docs/blueprints/EP035.rst

```diff
@@ -0,0 +1,208 @@
+to receive information on what services they are affecting,
+there needs to be a two way communication channel between
+the service provider and the interruption source.
+The kytos event bus provides only one way communication,
+making it unsuitable for this task.
+
+#############
+Specification
+#############
+
+The following is a set of proposed changes for
+how to handle interruptions in the future,
+in order to have visibility on there causes and effects.
+
+Kytos Core Changes
+##################
+
+In order for services to identify which interruptions are affecting the devices
+they use, a mechanism needs to be added to the device representation for
+identify interruption causes.
+
+
+Extend ``status_funcs`` to produce a ``status_reason``
+------------------------------------------------------
+
+Currently, when an object is queried about its ``status``,
+the reason for that status is unknown.
+What we propose here is adding a new property to
+``GenericEntity`` called ``status_reason``,
+and adding in ``status_reason_func``
+to produce values for ``status_reason``.
+The produced ``status_reason`` will be a set of strings,
+so all possible reasons a device might be disable
+can be easily represented.
+
+Example usage:
+
+.. code-block:: python3
+
+    from kytos.core.switch import Switch
+    maintenance_switches = set()
+
+    def switch_maintenance_status(switch):
+        if switch.id in maintenance_switches:
+            return EntityStatus.DOWN
+        return EntityStatus.UP
+
+    Switch.register_status_func('maintenance', switch_maintenance_status)
+
+    def switch_maintenance_reason(switch):
+        if switch.id in maintenance_switches:
+            return {'maintenance'}
+        return set()
+
+    Switch.register_status_reason_func('maintenance', switch_maintenance_reason)
+
+
+New Topology REST Endpoints
+###########################
+
+The following set of new endpoints are intended to allow
+for previewing the affects of adding in interruptions.
+
+
+REST Endpoint ``GET /topology/v3/preview_interruption``
+-------------------------------------------------------
+
+For all know service providers,
+previews affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: http
+
+    GET /topology/v3/preview_interruption?switch=<switch_id>&link=<link_id>&interface=<interface_id>&type=<interruption_type>
+
+
+Expected Response:
+
+.. code-block:: json
+
+    {
+        <provider name str>: [<service name str>],
+    }
+
+Side Effects:
+
+- None
+
+Notes:
+
+- Should support multiple key-value pairs with the same key in the query parameters.
+
+
+Endpoint: ``GET <url for previewing>``
+--------------------------------------
+
+For the given service provider,
+previews services affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: http
+
+    GET <preview_url>?switch=<switch_id>&link=<link_id>&interface=<interface_id>&type=<interruption_type>
```

Comment:
To simplify the query and filtering code on each provider, since each switch and link can be expressed as a function of interfaces, can we standardize to take a list of `interface` ids? 

Also, we might need to also specify another parameter to check if it's trying to check the interruption of an interface in a primary path or in a backup/failover path (not every service will have a concept of backup/faiolver path but for mef_eline it does). In this approach the caller client, topology, would need to send two requests and then also compare if impacted services had any backups or not, consolidating the response. Wdyt or how were you considering to address this part?

Reply:
A potential solution is to separate paths from services. Services specify which paths they may use, and interruptions query for which paths they affect.

Reply:
Yes. Sounds good to me, the client could parametrize which path like `?path_type="primary|backup"`. Trying to standardize this makes a lot of sense, and then it could be requested concurrently. 

Good thing that `mef_eline` evcs has the paths very well defined, so querying for that nested document object shouldn't be that difficult.